package_policies/R10.md

```diff
@@ -0,0 +1,65 @@
+**R10.** Each package should provide pre-processor macros that allow for version
+comparison (for languages that support it - like C/C++/Fortran).
+Using these macros, dependent software can potentially be compatible
+with multiple versions of the package.
+
+Example implementations for a given `package version 1.2.3`
+
+- Example 1: provide version macros only, via public include files (or can be done easily via CMake)
+```C
+#define <PACKAGE>_VERSION_MAJOR 1
+#define <PACKAGE>_VERSION_MINOR 2
+#define <PACKAGE>_VERSION_PATCH 3
+```
+
+- Example 2: provide  a single integer version macro to encode major, minor, patch version
+```C
+#define <PACKAGE>_VERSION 1002003
+```
+Additional comparison macros can help usage
+
+- With Example 1 (via public include files)
+```C
+#define <PACKAGE>_VERSION_GE(Major, Minor, Patch) \
+(((Major == <PACKAGE>_VERSION_MAJOR) && (Minor == <PACKAGE>_VERSION_MINOR) && (Patch <= <PACKAGE>_VERSION_PATCH)) || \
+        ((Major == <PACKAGE>_VERSION_MAJOR) && (Minor < <PACKAGE>_VERSION_MINOR)) || (Major < <PACKAGE>_VERSION_MAJOR))
+
+#define <PACKAGE>_VERSION_GT(Major, Minor, Patch) \
+(((Major == <PACKAGE>_VERSION_MAJOR) && (Minor == <PACKAGE>_VERSION_MINOR) && (Patch < <PACKAGE>_VERSION_PATCH)) || \
+        ((Major == <PACKAGE>_VERSION_MAJOR) && (Minor < <PACKAGE>_VERSION_MINOR)) || (Major < <PACKAGE>_VERSION_MAJOR))
+
+#define <PACKAGE>_VERSION_EQ(Major, Minor, Patch) \
+((Major == <PACKAGE>_VERSION_MAJOR) && (Minor == <PACKAGE>_VERSION_MINOR) && (Patch == <PACKAGE>_VERSION_PATCH))
+
+#define <PACKAGE>_VERSION_LE(Major, Minor, Patch) \
+!<PACKAGE>_VERSION_GT(Major, Minor, Patch)
+
+#define <PACKAGE>_VERSION_LT(Major, Minor, Patch) \
+!<PACKAGE>_VERSION_GE(Major, Minor, Patch)
+```
+
+- With Example 2 (via public include files)
+
+```C
+#define <PACKAGE>_VERSION_GE(Major, Minor, Patch) \
+(Major * 10000 + Minor * 100 + Patch <= <PACKAGE>_VERSION)
+
+#define <PACKAGE>_VERSION_GT(Major, Minor, Patch) \
+(Major * 10000 + Minor * 100 + Patch < <PACKAGE>_VERSION)
+
+#define <PACKAGE>_VERSION_LE(Major, Minor, Patch) \
+!<PACKAGE>_VERSION_GT(Major, Minor, Patch)
+
+#define <PACKAGE>_VERSION_LT(Major, Minor, Patch) \
+!<PACKAGE>_VERSION_GE(Major, Minor, Patch)
+```
+
+Implementation that are *NOT* acceptable include:
+
+- providing a string #define <PACKAGE>_VERSION "1.2.3"
+
+- expecting the user to "know" the version or expecting the user to manually edit source-code
+
+- requiring the user to use external tools such as CMake and git to generate and access version macros
```

Comment:
It's shorter but I don't know if it conveys what we want.

For example, in Tasmanian, the user must use CMake (or CMake though spack) to build and install Tasmanian and that's how the macros are generated. However, the macros are in the installed headers so the user does NOT need CMake to use Tasmanian or the macros.

Some people may misinterpret this and think that we forbid packages to use CMake to generate macros. But we want to avoid restrictions only on the USER, the package can generate the macors anyway they want.

Reply:
I pushed an update to clarify further